### PR TITLE
feat: integrate asyncapi-validator

### LINKED ISSUE #8: Integrate asyncapi-validator in Nodejs templates
@derberg commented on [Mon Mar 02 2020](https://github.com/asyncapi/generator/issues/234)

Reason:
Current Node templates nicely show what you can get from AsyncAPI file. You can generate an application that can work with Kafka or RabbitMQ, or others. When I play with those, there is one thing missing to make the templates nice, the message validation. Now you can send any message to the channel used by your application. Would be awesome to show that thanks to AsyncAPI with support of asyncapi-validator you can change that.

Description:
- Integrate https://www.npmjs.com/package/asyncapi-validator with the template
- Update docs and scratch note for release notes
- I'm wondering if it should be enabled by default or behind a parameter?

Something interesting for you @WaleedAshraf 😉 

---

@WaleedAshraf commented on [Mon Mar 02 2020](https://github.com/asyncapi/generator/issues/234#issuecomment-593330572)

🎉 Niceeee
I'd love to work on this, will try to join the next meeting and we can go from there. 👍 

---

@derberg commented on [Mon Mar 02 2020](https://github.com/asyncapi/generator/issues/234#issuecomment-593360366)

See ya tomorrow then!

### PR DESCRIPTION
**Pull request by KhudaDad414**
**Description**

-basic implementation added.

**Related issue(s)**
Resolves #8

### CODE REVIEW

**Comment by KhudaDad414** [comment]
@derberg @WaleedAshraf 
I want to add the validator as middleware.
my problem is msgIdentifier(that I have hard-coded for now.)
is making the name of the msg as msgIdentifier a good idea?
if it is, how can I access its value from message-validator.js? should I add the asyncapi-parser dependency as well?
any help would be appreciated.

**Comment by WaleedAshraf** [comment]
@KhudaDad414 Yes. Using the name is fine. It's a machine-friendly name for the message. see: https://asyncapi.io/docs/specifications/2.0.0#messageObject

This template is used with the generator: https://github.com/asyncapi/generator
generator already has the parser installed. Somehow you need to pass `message.name()` from generator to template. Just like `message` and `topic` are being passed.

File: template/src/api/middlewares/message-validator.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const AsyncApiValidator = require('asyncapi-validator');
+
+module.exports = (message, next) => {
+  if (message.headers.cmd === 'publish') {
+    validateMessage(message.payload, message.topic);
```
Comment (WaleedAshraf) [inline template/src/api/middlewares/message-validator.js:5]: I don't `message.topic` exists.
See these to check what properties exists: https://github.com/asyncapi/parser-js/blob/master/API.md#modules

You need `message.name`, `message.payload` and `channel` and `operation`.

**Comment by derberg** [comment]
@KhudaDad414 as @WaleedAshraf noted, this template is used with generator that gives you want you need. This template is running on older render engine so not React but Nunjucks.

- overall docs https://github.com/asyncapi/generator/blob/master/docs/authoring.md
- example template https://github.com/asyncapi/template-for-generator-templates/tree/nunjucks

Just look at this file from this template: https://github.com/asyncapi/nodejs-template/blob/master/template/src/api/index.js
The context of Nunjucks template has access to AsyncAPI document that is already parsed by the JS parser
```
{%- for channelName, channel in asyncapi.channels() %}
```

Complete API of AsyncAPI document after JS parsing -> https://github.com/asyncapi/parser-js/blob/master/API.md

**Pull request closed without merging**